File: python/example10_5.py
# Example 10.5 Higher Order Autocorrelation

import numpy as np
import pandas as pd
import statsmodels.api as sm

data = pd.read_csv('http://web.pdx.edu/~crkl/ceR/data/cjx.txt', sep='\s+', nrows=39)

X = np.log(data.X)
L = np.log(data.L1)
K = np.log(data.K1)
Z = pd.concat([L,K],axis=1)
Z = sm.add_constant(Z)
model_ols = sm.OLS(X,Z).fit()

def ols_ar(model,rho):
    x = model.model.exog
    y = model.model.endog
    p = rho.shape[0]
    ystar = y; xstar = x
    for i in range(1,p+1):
        ystar = ystar[1:]-rho[i-1]*y[:-i]
        xstar = xstar[1:,]-rho[i-1]*x[:-i,]
    model_ar = sm.OLS(ystar,xstar).fit()
    return(model_ar)

def OLSAR(model,order=1):
    x = model.model.exog
    y = model.model.endog
    e = (y - x @ model.params).reshape(-1,1)
    p = order
    e1 = e
    for i in range(1,p+1):
        e1 = np.hstack((e1[1:,],e[:-i]))
    rho0 = sm.OLS(e1[:,0],e1[:,1:]).fit().params
    print('Rho = ', rho0)
    rdiff = 1.0
    while(rdiff>1.0e-5):
        model1 = ols_ar(model,rho0)
        e = (y -x @ model1.params).reshape(-1,1)
        e1 = e
        for i in range(1,p+1):
            e1 = np.hstack((e1[1:,],e[:-i]))
        rho1 = sm.OLS(e1[:,0],e1[:,1:]).fit().params
        # rho1 = sm.OLS(e0,e1).fit().params
        rdiff = np.sqrt(np.sum((rho1-rho0)**2))
        rho0 = rho1
        print('Rho = ', rho0)
    return(ols_ar(model,rho0))

model_ar = OLSAR(model_ols,3)
model_ar.summary()

# check for residual autocorrelation of the final model
e = model_ar.resid
lags = 5
bptest = sm.stats.diagnostic.acorr_ljungbox(e, lags=lags, boxpierce=True, return_df=True)
print(bptest)

from statsmodels.graphics.tsaplots import plot_acf, plot_pacf
plot_acf(e,zero=False)
plot_pacf(e,zero=False)

# alternatively, using statsmodels' GLSAR
# estimate AR(1), AR(2), AR(3), 
# results may not be the same as OLSAR, need to check
for i in range(3):
    rho = i+1
    model_gls = sm.GLSAR(X,Z,rho)
    results = model_gls.iterative_fit(maxiter=50)
    print ('Iterations used = %d Converged %s' % (results.iter, results.converged) )
    print ('Rho =  ', model_gls.rho)
    print(results.summary())

# alternatively, ARIMA model may be used
# results may not be the same as OLSAR and GLSAR, need to check
from statsmodels.tsa.arima_model import ARMA
for i in range(3):
    rho = i+1
    armodel = ARMA(X,exog=Z,order=(rho,0)).fit(trend='nc')
    print(armodel.summary())

File: R/example10_5.R
# Example 10.5 Higher Order Autocorrelation

setwd("C:/Course17/ceR/R")
data<-read.table("http://web.pdx.edu/~crkl/ceR/data/cjx.txt",header=T,nrows=39)

year <- data$YEAR
X <- log(data$X)
L <- log(data$L1)
K <- log(data$K1)
results.ols<-lm(X~L+K)
summary(results.ols)

# cochrane-orcutt lm, given AR(p) rho, derived from ols model
lm.ar <- function(model,rho) {
  x <- model.matrix(model)
  y <- model.response(model.frame(model))
  r <- c(1,-rho)
  ystar<-filter(y,r,sides=1)
  xstar<-filter(x,r,sides=1)
  return(lm(ystar~xstar-1))
}

# cochrane-orcutt procedure
lm.co<-function(model,order) {
  x<-model.matrix(model)
  y<-model.response(model.frame(model))
  e<-y-x%*%model$coefficients
  p<-order+1
  estar<-embed(e,p)
  r0<-lm(estar[,1]~estar[,2:p]-1)$coefficients
  rdiff<-1
  while(rdiff>1.0e-5) {
    model1<-lm.ar(model,r0)
    e<-y-x%*%model1$coefficients
    estar<-embed(e,p)
    rho<-lm(estar[,1]~estar[,2:p]-1)$coefficients
    rdiff<-sum((rho-r0)^2)
    r0<-rho
  }
  list(lm.co=summary(lm.ar(model,r0)),rho=r0)
}
results.co<-lm.co(results.ols,3) # the results are based on transformed model
results.co

# check for residual autocorrelation
e<-results.co[[1]]$residuals
for(i in 1:4) print(Box.test(e,lag=i,type="Box-Pierce"))
for(i in 1:4) print(Box.test(e,lag=i,type="Ljung-Box"))

acf(e,type="correlation")
acf(e,type="partial")  # same pacf(results$resid)

# alternatively using package nlme
library(nlme)
results.gls<-gls(X~L+K,correlation=corARMA(p=3), method="ML")
summary(results.gls)

rm(list=ls())
